### DISCUSSION TITLE
How do you perform partial searches in pgroonga?

### DISCUSSION THREAD

**Discussion by GabenGar** [pgroonga/pgroonga#248] [Q&A] [answered]
Given the table:
```sql
CREATE TABLE people (
  id bigint GENERATED ALWAYS AS IDENTITY PRIMARY KEY,
  name text NOT NULL
);

CREATE INDEX pgroonga_content_index ON people
  USING pgroonga (name);
```
The values:
```sql
INSERT INTO people
  (
    name
  )
VALUES
  ('Jack Yack'),
  ('Ackerman')
;
```
I would expect the query:
```sql
SELECT
  name
FROM
  people
WHERE
  name &@* 'ack'
;
```
Also return `Jack Yack` in the results, but it doesn't happen.

**Comment by komainu8** [comment, marked as answer]
Could you create index with `TokenNgram` as below.

```sql
CREATE INDEX pgroonga_content_index ON people
  USING pgroonga (name) WITH(tokenizer='TokenNgram("unify_alphabet", false, "unify_symbol", false, "unify_digit", false)');
```

In addition `&@*` is performs similar search ( `&@*` is not the operator for full text search).
We should use `&@~` operator in this case as below.

```sql
SELECT
  name
FROM
  people
WHERE
  name &@~ 'ack'
;
   name    
-----------
 Jack Yack
 Ackerman
(2 rows)
```

**Reply by GabenGar** [reply]
Is it a good idea to disable all those options in the index? The example data was english-only for simplicity sake, the actual data is going to be a multi-language mix with all sorts of weird characters, which isn't even guarantied to be collatable to a BCP-47 tag within a single string in the first place. And these options sound like they would disable a lot of useful features.

**Reply by komainu8** [reply]
If those option disable, PGroonga uses bigram tokenize method for alphabet, symbol, and digits.

Bigram tokenize method is good for recall because you can find all texts by query consists of two or more characters.
However, bigram tokenize method isn’t good for precision because you can find texts that includes query in word. For example, you can find world by or. 

If we value recall over precision in your usecase, we should these option disable.
